BUG: Strings with exponent but no decimal point parsed as integers in python csv engine

evanpw opened the pull request:
Fixes GH #9565. I also made the handling of out-of-range integers consistent with the C engine: parse as a string rather than a double.

evanpw committed:
BUG: Strings with exponent but no decimal point parsed as integers in python csv engine (GH 9565)

jreback commented on pandas/src/inference.pyx:
```diff
@@ -569,18 +569,17 @@ def maybe_convert_numeric(object[:] values, set na_values,
+                        if fval < fINT64_MAX and fval > fINT64_MIN:
+                            try:
+                                ints[i] = int(val)
+                            except ValueError:
```
I believe this can still fail for an int > int64max (but less < than a uint64max)

evanpw replied:
It produces a string in that case:
``` .python
In [4]: pd.read_csv(StringIO(str(np.iinfo(np.int64).max + 1)), header=None, engine='python').values
Out[4]: array([['9223372036854775808']], dtype=object)

In [5]: pd.read_csv(StringIO(str(np.iinfo(np.int64).min - 1)), header=None, engine='python').values
Out[5]: array([['-9223372036854775809']], dtype=object)
```

Do you want it to infer the dtype ``uint64``? The C parser doesn't do that either currently.

I don't see how that line ``ints[i] = int(val)`` could ever throw an exception, though. Maybe there was some case that's now been eliminated?

jreback commented:
pls give this a perf run on the ``-r csv`` and see if anything is changed

jreback commented:
@evanpw looks good. (FYI if you want to add a confirming case for the out-of-range int64 you have above would be good - not sure that is really tested atm).

and pls run perf and report if anything amiss.

evanpw commented:
Integers near the boundary of int64 values were getting parsed as strings because the boundary comparison was performed using doubles. That's fixed in this last commit, and some tests added. Checked performance: no measurable difference with master. Rebased and moved whatsnew entry to 0.16.2.

jreback commented:
fyi, you may have to rebase again as I just fixed a travis build error about 10 min ago.

jreback commented on pandas/lib.pyx:
```diff
@@ -64,11 +64,7 @@ ctypedef unsigned char UChar
-    enum: INT64_MAX
-    enum: INT64_MIN
-
+from libc.stdint cimport UINT8_MAX, INT64_MAX, INT64_MIN
```
this doesn't work on windows, just revert this section

jreback commented on pandas/parser.pyx:
```diff
@@ -45,19 +45,7 @@ cdef bint PY3 = (sys.version_info[0] >= 3)
-    enum: INT32_MIN
-    enum: INT64_MAX
-    enum: INT64_MIN
+from libc.stdint cimport *
```
same here

jreback commented on pandas/src/inference.pyx:
```diff
@@ -514,10 +514,7 @@ def is_period_array(ndarray[object] values):
-    inline int floatify(object, double *result) except -1
-
-cdef double fINT64_MAX = <double> INT64_MAX
-cdef double fINT64_MIN = <double> INT64_MIN
```
same here (with the definitions)

jreback commented:
couple of issues w.r.t. windows. I think some of the libraries are named differently. So just revert that part of the code. ping after travis is green.

evanpw commented:
cython libc imports reverted, and tests are green

jreback commented:
merged via https://github.com/pydata/pandas/commit/38876df03ada8e962fad06ac64cb343562cb976c

thanks!
